### ISSUE TITLE
EventSource and Sensor for multiple Github Repos

### ISSUE THREAD

**Issue by ryanli-me** [argoproj/argo-events#976]
# Summary

What do you want to know about this project?

What's the best way to have github event source and sensor for multiple Github Repos? With the help of the examples in this repo and those in argo workflow, I currently have a CI pipeline for a single github repo with Argo eventsource, sensor and workflow working together. However, what is the best practice to have the event source listen on multiple github repo's update while sensor could run different workflow depending on the repo's specifications (languages)? I'm able to get the parameters parsed from event source (ref, repo ssh url) and passed to workflow. I have two approaches in mind to have the pipeline listen on multiple Github repos and wonder if they are feasible and what's the best practice?

1. Have one single github EventSource with multiple event each pointing to a different Github repo that I want to manage, they all have the same webhook url which is an ingress pointing to `github-eventsource-svc`. Then have one single sensor to parse the info from webhook payload and create workflow. However, this approach requires a single file to manage all the different repos my org has which is not ideal. I'd prefer to have this EventSource config live in each individual repos. Also, I'm not sure if I could I create only one single sensor that depends on github event source without specifying event name to sub on all different repo's webhook payload.

2. Create individual EventSource and Sensor in every repo that requires a CI pipeline. This could lead to a lot of boilerplate which is hard to manage and update. Also I believe each EventSource object will lead to a service and a deployment which could lead to a lot of resources consumed on Kubernetes cluster. The good part is I could have ArgoCD to manage EventSource and Sensor for individual repo along side the other YAMLs for those repos.

# Motivation

Why do you need to know this, any examples or use cases you could include?

A good way to build a CI pipeline for multiple repos.

**Comment by whynowy** [comment]
First of all, both approaches are feasible.

For approach 1, single sensor is doable by specifying [Trigger Conditions](https://github.com/argoproj/argo-events/blob/master/docs/trigger-conditions.md).

For approach 2, even you do one EventSource per repo, you still can use ingress to aggregate them into one webhook url. In terms of resource consumption, both EventSource and Sensor consume very limited CPU and memory.

Can not tell which approach is better, it really depends on what kind of system you want to build. Or you can take the middle, that do kind of aggregation (e.g. aggregate repos from same team), so that you don't have many files, also don't have many PODs.

**Comment by sjbodzo** [comment]
I found this issue while trying to solve a similar problem. I am going to list out some of the things I tried, in case anyone else benefits from sharing.

TLDR: To use all repos in a Github org, just manually create a webhook trigger and point it to the ingress for your Argo Event. Then, use the generic webhook trigger type on the `EventSource` you create.

### Attempt 1
Since our YAML is Helm templated, use a Python script to fetch and output a yaml list, then range over the list for the _single_ `EventSource` object and the `dependencies` section of the `Sensor` object.

Problems: Argo encodes some data into both the sensor and event source pods as environment variables. Unfortunately, the data of these environment variables is directly related to the repositories configured. For us, roughly 200 repos in our list maxed out the environment variable data. Anything more than that causes the pod to repeatedly crash.

### Attempt 2
Manually create an org-level webhook in Github and register a single repository in that org as the `EventSource`. The thinking here was that the ingress created in the cluster would be hit by all repositories in the org, and Argo would be "tricked". 

Problems: In testing it appears that the Event Bus acknowledges the events, but then Argo filters them out rather than triggering the `Sensor` (since the repositories do not match the listed repos on the `dependencies`). 

### Attempt 3
Manually create an org-level webhook in Github, and use a generic webhook trigger for the `EventSource`. This one works, with the caveat that you lose the automation per-repo webhook creation that Argo does for you with the Github trigger type.

**Comment by Hunter-Thompson** [comment]
Could something like the following be implemented?

```
spec:
  github:
    docker_tag:
      # owner of the repo
      owner: example
      # repository names comma delimited 
      repository: helloworld-python, helloworld-python-2, helloworld-python-3
 ```
A comma delimited list of repos the event source will be listening on.

**Comment by whynowy** [comment]
Please comment on https://github.com/argoproj/argo-events/issues/1112 @ryanli-me

**Comment by whynowy** [comment]
This has been implemented.

**Comment by dave08** [comment]
In your answer @whynowy, you said:

> For approach 2, even you do one EventSource per repo, you still can use ingress to aggregate them into one webhook url. In terms of resource consumption, both EventSource and Sensor consume very limited CPU and memory.

How can I aggregate them to one webhook? I'm having this problem with the bitbucket cloud webhook, and as far as I understand a k8s service round-robins between pods -- so if each eventsource is a different pod then the ingress will send to the service with will randomly choose one eventsource to send to -- not necessarily the right one though...
